packages/upload/src/vaadin-upload-mixin.js

```diff
@@ -558,12 +558,49 @@ export const UploadMixin = (superClass) =>
     }
 
     /** @private */
-    _onDrop(event) {
+    async _onDrop(event) {
       if (!this.nodrop) {
         event.preventDefault();
         this._dragover = this._dragoverValid = false;
-        this._addFiles(event.dataTransfer.files);
+
+        const files = await this.__getFilesFromDropEvent(event);
+        this._addFiles(files);
+      }
+    }
+
+    /**
+     * Get the files from the drop event. The dropped items may contain a
+     * combination of files and directories. If a dropped item is a directory,
+     * it will be recursively traversed to get all files.
+     *
+     * @param {!DragEvent} dropEvent - The drop event
+     * @returns {!Array<!File>} - The files from the drop event
+     * @private
+     */
+    async __getFilesFromDropEvent(dropEvent) {
+      async function getFilesFromEntry(entry) {
+        if (entry.isFile) {
+          return new Promise((resolve) => {
+            // In case of an error, resolve without any files
+            entry.file(resolve, () => resolve([]));
+          });
+        } else if (entry.isDirectory) {
+          const reader = entry.createReader();
+          const entries = await new Promise((resolve) => {
+            // In case of an error, resolve without any files
+            reader.readEntries(resolve, () => resolve([]));
+          });
+          const files = await Promise.all(entries.map(getFilesFromEntry));
+          return files.flat();
+        }
       }
+
+      const filePromises = Array.from(dropEvent.dataTransfer.items)
+        .map((item) => item.webkitGetAsEntry())
```

Comment:
At the moment [webkitGetAsEntry](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry) seems to be the only API that allows retrieving a folder from a data transfer item that is supported by all browsers. 

There is also [getAsFileSystemHandle](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/getAsFileSystemHandle), but that isn't supported by FF or Safari. Then there are some mentions of other APIs in the referenced issue, none of which seem to have made it into the standard.